[FIX] project: check company

HANNICHE-Walid opened the pull request:
Steps to reproduce:
- Install Stock and Project
- Create a new project with current company as customer (3 dots menu on project overview to set the customer)
- Create a new warehouse for the current company
- Error pops

Bug:
since commit[1] check company is enforced on the project_id field of partners in bothways
when creating a new warehouse CheckCompany fields are editted on the partner if set (property_stock_customer/supplier) the issue is if the partner's company_id is False while project belongs to company "_check_company" fails

Fix:
Relax condition on check company to allow for company False to pass for corecords

alternatives:
- Set partner's company to the project company when adding it as a customer
- Add a domain to only allow partners that belongs to the current company
- Remove check company on project_ids from the partner's side

opw-3507952
[1]:https://github.com/odoo/odoo/pull/109464

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/138398).

HANNICHE-Walid commented:
also solves opw-3505050

HANNICHE-Walid commented:
hi @rco-odoo can you take a look at this PR please

ryv-odoo commented:
Hello @HANNICHE-Walid , 

This change is not trivial and changes the behavior of `check_company`. In our technical documentation there is even a warning block for your case: https://www.odoo.com/documentation/16.0/developer/howtos/company.html#multi-company-consistency

Anyway, this kind of change in the ORM should be highly motivated and not done in a stable version. Then I advise to look for an alternative solution (maybe have a more relaxed domain on your fields)

Note that it is still strange that the behavior is not covered by any test.

HANNICHE-Walid commented:
hi @ryv-odoo the issue I have with check company right now is that
if the record has company False and corecord has CompanyA it passes
but if records has CompanyA and the corecord has company False if fails
that's why I proposed it should pass too in that case
I can retarget to master if that's the expected behaviour

HANNICHE-Walid commented:
@odoo/rd-services I have mentioned some alternatives if you can pick the most appropriate

xavierbol commented:
Hello @HANNICHE-Walid ,

You can remove check company on project_ids from the partner's side. 🙂

xavierbol commented:
Can you also write a test to cover the problematic case?

The one with partner without any company set and project with a company set and that partner set. 

Thanks in advance!

xavierbol commented on addons/project/models/res_partner.py:
```diff
@@ -8,9 +8,8 @@
-    _check_company_auto = True
 
-    project_ids = fields.One2many('project.project', 'partner_id', string='Projects', check_company=True)
+    project_ids = fields.One2many('project.project', 'partner_id', string='Projects')
```
Can we add a python contraint to avoid having a partner in company A and set in project in company B for instance? 🤔 

Because the `check_company` we added was really to avoid that. 🙂 

Can you check? 🙂 

Thanks in advance! 🙂

HANNICHE-Walid replied:
done Thanks for the feedback :)

HANNICHE-Walid commented:
hi @xavierbol I have added the requested changes

xavierbol reviewed (commented):
Thanks again for your work. 🙂 

In fact, I am wondering if the constraint we added if really needed since that part:
```py
        with self.assertRaises(UserError):
            # Cannot change the company of a partner if the company of the project is different
            partner.company_id = company_1
```
in `test_change_project_or_partner_company` test, raise an error as expected if we check the last build in runbot for your branch. 🤔 

And so, I think that constrains is a bit overkill then. Sorry about that... 😬 🙈

HANNICHE-Walid commented:
> Thanks again for your work. 🙂
> 
> In fact, I am wondering if the constraint we added if really needed since that part:
> 
> ```python
>         with self.assertRaises(UserError):
>             # Cannot change the company of a partner if the company of the project is different
>             partner.company_id = company_1
> ```
> 
> in `test_change_project_or_partner_company` test, raise an error as expected if we check the last build in runbot for your branch. 🤔
> 
> And so, I think that constrains is a bit overkill then. Sorry about that... 😬 🙈

The error is actually raised for the python constraint I had removed that part for the test to pass without it initially 
so it is required indeed

HANNICHE-Walid commented on addons/project/models/res_partner.py:
```diff
@@ -1,19 +1,25 @@
+    def _check_active(self):
+        for record in self:
+            if record.company_id and record.project_ids.company_id and record.project_ids.company_id != record.company_id:
+                raise UserError(_("Partner company can't be diffirent from its assigned project's company"))
```
better indeed :) thanks

xavierbol commented:
Oh okay. 😄

xavierbol reviewed (approved):
LGTM, thanks again for your work. 🙂 

robodoo r+

xavierbol reviewed (commented):
robodoo r-

Can you update the commit title since the changes are no longer in base module?

thanks in advance!

HANNICHE-Walid commented:
@xavierbol done
 thanks for the review <3

xavierbol commented:
robodoo r+

robodoo commented:
@xavierbol you may want to rebuild or fix this PR as it has failed CI.
